File: leetcode/AIcoding/1.Gemini-2.5/cpp_solutions/869-reordered-power-of-2.cpp
#include <vector>
#include <string>
#include <algorithm>
#include <cmath>

class Solution {
public:
    bool reorderedPowerOf2(int n) {
        std::vector<int> n_counts = countDigits(n);

        // Iterate through powers of 2.
        // n is up to 10^9.
        // 2^0 = 1
        // ...
        // 2^29 = 536,870,912 (9 digits)
        // 2^30 = 1,073,741,824 (10 digits)
        // The largest n is 10^9, which has 10 digits.
        // So we need to check powers of 2 that have up to 10 digits.
        // 2^30 is the largest power of 2 that can be formed from digits of a number <= 10^9.
        for (int i = 0; i <= 30; ++i) {
            int power_of_2 = 1 << i;
            std::vector<int> p2_counts = countDigits(power_of_2);
            if (n_counts == p2_counts) {
                return true;
            }
        }
        return false;
    }

private:
    // Helper function to count digit frequencies of a number
    std::vector<int> countDigits(int num) {
        std::vector<int> counts(10, 0);
        while (num > 0) {
            counts[num % 10]++;
            num /= 10;
        }
        return counts;
    }
};

File: leetcode/AIcoding/3.DeepSeek-v3/go_solutions/869-reordered-power-of-2.go
func reorderedPowerOf2(n int) bool {
    countDigits := func(num int) [10]int {
        count := [10]int{}
        for num > 0 {
            digit := num % 10
            count[digit]++
            num /= 10
        }
        return count
    }

    target := countDigits(n)
    for i := 0; i < 31; i++ {
        power := 1 << i
        current := countDigits(power)
        if current == target {
            return true
        }
    }
    return false
}